prow/github/client.go

```diff
@@ -185,18 +190,36 @@ func (c *Client) Throttle(hourlyTokens, burst int) {
 	c.throttle.throttle = throttle
 }
 
+// RoundTrip is an implementation for net/http RoundTripper interface.
+// Its being used to refresh the token from a token source.
+func (ct *cacherTransport) RoundTrip(req *http.Request) (resp *http.Response, err error) {
+	_, err = ct.Base.Source.Token()
+	if err != nil {
+		return nil, err
+	}
+	// ref: https://godoc.org/net/http#Transport.RoundTrip
+	resp, err = ct.Base.RoundTrip(req)
+	if err != nil {
+		return nil, err
+	}
+	_, err = ct.Base.Source.Token()
+	if err != nil {
+		return nil, err
+	}
+	return resp, nil
+}
+
 // NewClient creates a new fully operational GitHub client.
 // 'getToken' is a generator for the GitHub access token to use.
 // 'bases' is a variadic slice of endpoints to use in order of preference.
 //   An endpoint is used when all preceding endpoints have returned a conn err.
 //   This should be used when using the ghproxy GitHub proxy cache to allow
 //   this client to bypass the cache if it is temporarily unavailable.
-func NewClient(getToken func() []byte, bases ...string) *Client {
+func NewClient(getToken func() []byte, tokenSource oauth2.TokenSource, bases ...string) *Client {
 	return &Client{
-		logger: logrus.WithField("client", "github"),
-		time:   &standardTime{},
-		gqlc: githubql.NewClient(oauth2.NewClient(context.Background(),
-			oauth2.StaticTokenSource(&oauth2.Token{AccessToken: string(getToken())}))),
+		logger:   logrus.WithField("client", "github"),
+		time:     &standardTime{},
+		gqlc:     githubql.NewClient(&http.Client{Transport: &cacherTransport{Base: &oauth2.Transport{Source: tokenSource}}}),
```

Comment:
There is an issue in the oauth2 library, that it can't pass the context to different levels. So the ReuseTokenSource is not generating the token from the source. That's why the roundtrip is needed.
See ref: [oauth2 issue 84](https://github.com/golang/oauth2/issues/84)